### PR TITLE
[Feature] idXML multiple evidence support

### PR DESCRIPTION
**Pull request by timosachsenberg**
this PR adds support for multiple peptide evidences in idXML.
Due to the large number of changes I decided to do several incremental PRs (first idXML, than consensusXML and featureXML)

Code is read-backwards compatible but a schema change was required.

aa_before and aa_after now store list of characters
start and end list of positions
together with the list of protein references valuable information for: e.g. protein converage calculation can now carried over between idXML files.

### CODE REVIEW

File: src/tests/topp/FalseDiscoveryRate_output_4.idXML [hunk 1]
```diff
@@ -2179,7 +2179,7 @@
 			</PeptideHit>
 		</PeptideIdentification>
 		<PeptideIdentification score_type="q-value" higher_score_better="false" significance_threshold="0" MZ="539.28" RT="990" >
-			<PeptideHit score="0.121212121212121" sequence="LLDEVEELLEQER" charge="3" aa_before="R" aa_after="L" protein_refs="PH_9 PH_10 PH_11 PH_12 PH_464" >
+			<PeptideHit score="0.121212121212121" sequence="LLDEVEELLEQER" charge="3" aa_before="R X X X X" aa_after="L X X X X" protein_refs="PH_9 PH_10 PH_11 PH_12 PH_464" >
```
Comment (hroest) [inline src/tests/topp/FalseDiscoveryRate_output_4.idXML]: why are all the other `aa_before` and `aa_after` denoted as "X", e.g. unknown? Why is only the first one known?

File: src/openms/include/OpenMS/FORMAT/IdXMLFile.h [hunk 2]
```diff
@@ -99,6 +99,19 @@ namespace OpenMS
         @exception Exception::UnableToCreateFile is thrown if the file could not be created
     */
     void store(String filename, const std::vector<ProteinIdentification>& protein_ids, const std::vector<PeptideIdentification>& peptide_ids, const String& document_id = "");
+  
+    /**
+      * Helper function to create the XML string for the amino acids before and after the peptide position in a protein.
+      * Can be eused by e.g. ConsensusXML, FeatureXML to write PeptideHit elements  
+      */
+    static String createFlankingAAXMLString(const std::vector<PeptideEvidence> & pes);
```
Comment (hroest) [inline src/openms/include/OpenMS/FORMAT/IdXMLFile.h]: could you add these two fxn to the Python bindings?

File: src/tests/topp/FalseDiscoveryRate_output_4.idXML (hunk 1 above)
Comment (timosachsenberg) [inline src/tests/topp/FalseDiscoveryRate_output_4.idXML]: That is how it was used before. Only for the first peptide evidence (first match to a protein) this information was available in idXML (for whatever reason). Newer idXML should contain this information for every hit. I adapted the TandemXML file parser but full support will probably be added step by step.

File: src/openms/include/OpenMS/FORMAT/IdXMLFile.h (hunk 2 above)
Comment (timosachsenberg) [inline src/openms/include/OpenMS/FORMAT/IdXMLFile.h]: They are not really helpful but I will add them to make tests pass.

Comment (timosachsenberg) [inline src/openms/include/OpenMS/FORMAT/IdXMLFile.h]: I made them protected and added ConsensusXMLFile and FeatureXMLFile as friend classes to prevent code duplication

**Comment by timosachsenberg** [comment]
related to: https://github.com/OpenMS/OpenMS/issues/1031

**Comment by timosachsenberg** [comment]
Travis passed

**Comment by timosachsenberg** [comment]
if someone finds time to review this PR I would be great as this type of PR looses mergeability pretty fast.
Changes are not that large (except many adapted test files)

**Comment by hendrikweisser** [comment]
Can you represent cases where a peptide matches several times in the same protein, with different neighbouring amino acids?

File: src/openms/include/OpenMS/FORMAT/IdXMLFile.h [hunk 3]
```diff
@@ -65,6 +65,9 @@ namespace OpenMS
     public Internal::XMLFile
   {
 public:
+    // both ConsensusXMLFile and FeatureXMLFile use some protected IdXML helpfer functions to parse identifications without code duplication
```
Comment (jpfeuffer) [inline src/openms/include/OpenMS/FORMAT/IdXMLFile.h]: *helper

File: src/openms/include/OpenMS/FORMAT/IdXMLFile.h [hunk 4]
```diff
@@ -113,6 +117,18 @@ namespace OpenMS
     /// Read and store ProteinGroup data
     void getProteinGroups_(std::vector<ProteinIdentification::ProteinGroup>& groups, const String& group_name);
 
+    /**
+      * Helper function to create the XML string for the amino acids before and after the peptide position in a protein.
+      * Can be eused by e.g. ConsensusXML, FeatureXML to write PeptideHit elements
```
Comment (jpfeuffer) [inline src/openms/include/OpenMS/FORMAT/IdXMLFile.h]: *reused
Also in the following copies of the comment.

File: src/openms/include/OpenMS/METADATA/PeptideEvidence.h [hunk 5]
```diff
@@ -55,7 +55,7 @@ namespace OpenMS
 
     // Note: we use 0 as position of the N-terminus while e.g. mzTab or other formats start counting at 1.
     static const int N_TERMINAL_POSITION;
-    static const char UNKNOWN_AA; // PeptideEvidence::UNKNOWN_AA = ' ';
+    static const char UNKNOWN_AA; // PeptideEvidence::UNKNOWN_AA = 'X';
```
Comment (jpfeuffer) [inline src/openms/include/OpenMS/METADATA/PeptideEvidence.h]: What is the reason for the change?
Just curious.

**Comment by timosachsenberg** [comment]
yes that's possible. In OpenMS you have two PeptideEvidence objects with the same protein accession but different positon/flanking amino acids. In idXML the aa_before and aa_after elements are now a list (like the protein accession list in PeptideHit)

File: src/openms/include/OpenMS/METADATA/PeptideEvidence.h (hunk 5 above)
Comment (timosachsenberg) [inline src/openms/include/OpenMS/METADATA/PeptideEvidence.h]: XML stores lists as space separated and we now need non-whitespace characters for the list entries.

**Comment by hendrikweisser** [comment]
@timosachsenberg:
>  In idXML the aa_before and aa_after elements are now a list (like the protein accession list in PeptideHit)

So in the case of multiple matches in the same protein, do you repeat the protein accession in the "protein_refs" element?

**Comment by timosachsenberg** [comment]
exactly

File: src/tests/topp/PeptideIndexer_2_out.idXML [hunk 6]
```diff
@@ -1,55 +1,55 @@
...
+				<UserParam type="string" name="target_decoy" value="decoy"/>
 			</ProteinHit>
 			<ProteinHit id="PH_8" accession="revonly_rev" score="0" sequence="DFIANGER" >
-                <UserParam type="string" name="target_decoy" value="decoy"/>
+				<UserParam type="string" name="target_decoy" value="decoy"/>
 			</ProteinHit>
 		</ProteinIdentification>
 		<PeptideIdentification score_type="Mascot" higher_score_better="true" significance_threshold="31.8621" MZ="0" RT="0" >
-			<PeptideHit score="40" sequence="STFDFYQRRLVTLAESPRAPSP" charge="2" aa_before="X" aa_after="G" protein_refs="PH_0 PH_2" >
+			<PeptideHit score="40" sequence="STFDFYQRRLVTLAESPRAPSP" charge="2" aa_after="G ]" start="227 4"  end="248 25" protein_refs="PH_0 PH_2" >
```
Comment (jpfeuffer) [inline src/tests/topp/PeptideIndexer_2_out.idXML]: Why is aa_before missing here?

**Comment by jpfeuffer** [comment]
Check my last comment, then +1.

File: src/tests/topp/PeptideIndexer_2_out.idXML (hunk 6 above)
Comment (timosachsenberg) [inline src/tests/topp/PeptideIndexer_2_out.idXML]: There is no aa_before information available (all are unknown) so writing is omitted.
Reading it back in produces the correct result.

**Comment by hendrikweisser** [comment]
@timosachsenberg:
I just noticed that you forgot to update the FidoAdapter test files (in the THIRDPARTY folder).
Also, can you check what's going on with `src/tests/topp/FeatureLinkerUnlabeledQT_2_output.consensusXML` - that test file has changed a lot, in parts that don't seem related to peptide evidence.

**Comment by timosachsenberg** [comment]
ok thanks I will update this - seems the order of entries is changed.

**Comment by timosachsenberg** [comment]
fido fixed in https://github.com/OpenMS/OpenMS/pull/1681
